### PR TITLE
[OODT-1010] Improve JAXRS REST APIs

### PR DESCRIPTION
**Pull request by NGimhana**
This PR related to JIRA Issue [OODT-1010](https://issues.apache.org/jira/browse/OODT-1010).

### CODE REVIEW

File: webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/resources/ProductPageResource.java [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+/**
+ * A JAX-RS resource representing a {@link ProductPage}.
+ * @author ngimhana (Nadeeshan Gimhana)
+ */
+@XmlRootElement(name="productPage")
+@XmlType(propOrder = {"pageSize","pageNum","totalPages","numOfHits","metadataResource", "productResources"})
+@XmlAccessorType(XmlAccessType.NONE)
+public class ProductPageResource
+{
```
Comment (IMS94) [inline webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/resources/ProductPageResource.java]: Formatting should be consistent. You can import Google Java format to IDEA from [this file](https://github.com/google/styleguide/blob/gh-pages/intellij-java-google-style.xml).

File: webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/resources/ProductPageResource.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+    private MetadataResource metadataResource;
+    private List<ProductResource> productResources = new ArrayList<ProductResource>();
+    private int pageSize;
+    private int pageNum;
+    private int totalPages;
+    private long numOfHits;
+
+
+    /* The file manager's working directory for this resource, used for example
```
Comment (IMS94) [inline webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/resources/ProductPageResource.java]: Should be a javadoc -> `/** <content> */`

File: webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/resources/ProductPageResource.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+        this.pageSize = page.getPageSize();
+        this.pageNum = page.getPageNum();
+        this.totalPages = page.getTotalPages();
+        this.numOfHits = page.getNumOfHits();
+
+        this.workingDir = workingDir;
+
+        List<Product> pageProducts = page.getPageProducts();
+        for(int i=0;i<pageProducts.size();i++){
```
Comment (IMS94) [inline webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/resources/ProductPageResource.java]: Should be formatted to have spaces in between.

File: webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/resources/ProductPageResource.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+            this.productResources.add(new ProductResource(product,metadata,productReferencesList.get(0),workingDir));
+        }
+
+
+    }
+
+
+
+
```
Comment (IMS94) [inline webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/resources/ProductPageResource.java]: Unnecessary new lines?

File: webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/services/CasProductJaxrsService.java [hunk 5]
```diff
@@ -40,10 +41,7 @@
 import java.util.logging.Logger;
 
 import javax.servlet.ServletContext;
-import javax.ws.rs.GET;
-import javax.ws.rs.Path;
-import javax.ws.rs.Produces;
-import javax.ws.rs.QueryParam;
+import javax.ws.rs.*;
```
Comment (IMS94) [inline webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/services/CasProductJaxrsService.java]: We shouldn't be using * imports. Importing a code formatter as mentioned in an earlier comment and reformatting these files will fix that. Or you will have to manually change settings in IDEA.

File: webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/services/CasProductJaxrsService.java [hunk 6]
```diff
@@ -361,9 +357,11 @@ public FileManagerClient getContextClient()
...
     }
 
-    String message = "Unable to get the file manager client from the "
-      + "servlet context.";
+    String message = ErrorTypes.CAS_PRODUCT_EXCEPTION_FILEMGR_CLIENT_UNAVILABLE.getErrorType();
     LOGGER.log(Level.WARNING, message);
     throw new CasProductException(message);
   }
+
+
```
Comment (IMS94) [inline webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/services/CasProductJaxrsService.java]: Formatting 😄 ...

File: webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/services/FileManagerJaxrsServiceV2.java [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * Service class for Proposing Apache OODT-2.0 FileManager REST-APIs
+ * This handles HTTP requests and returns file manager entities
+ * JAX-RS resources converted to different formats.
+ * @author ngimhana (Nadeeshan Gimhana)
+ */
+public class FileManagerJaxrsServiceV2 {
```
Comment (IMS94) [inline webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/services/FileManagerJaxrsServiceV2.java]: Entire class require to be formatted properly

File: webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/services/FileManagerJaxrsServiceV2.java [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+            // Create the product resource, add the product data and return the
+            // resource as the HTTP response.
+            return new ProductResource(product, client.getMetadata(product),
+                    product.getProductReferences(), getContextWorkingDir());
+        }
+        catch (Exception e)
+        {
+            // Just for Logging Purposes
+            String message = "Unable to find the requested resource.";
+            LOGGER.log(Level.FINE, message, e);
```
Comment (IMS94) [inline webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/services/FileManagerJaxrsServiceV2.java]: It will be ok to print an error log. Fine for now.

File: webapp/fmprod/src/main/webapp/WEB-INF/web.xml [hunk 9]
```diff
@@ -72,7 +72,7 @@
     <init-param>
       <param-name>jaxrs.serviceClasses</param-name>
       <param-value>
-        org.apache.oodt.cas.product.jaxrs.services.CasProductJaxrsService
```
Comment (IMS94) [inline webapp/fmprod/src/main/webapp/WEB-INF/web.xml]: Just to clarify. This means we will not be using `CasProductJaxrsService` in the future and it will be replaced by `FileManagerJaxrsServiceV2`?

File: webapp/fmprod/src/test/java/org/apache/oodt/cas/product/jaxrs/resources/DatasetResourceTest.java [hunk 10]
```diff
@@ -70,7 +71,7 @@ public void testXmlMarshalling() throws IOException, JAXBException,
     List<Reference> references1 = new ArrayList<Reference>();
     references1.add(reference1);
 
-    Hashtable metadataEntries1 = new Hashtable<String, Object>();
+    HashMap<String, Object> metadataEntries1 = new HashMap<>();
```
Comment (IMS94) [inline webapp/fmprod/src/test/java/org/apache/oodt/cas/product/jaxrs/resources/DatasetResourceTest.java]: For clarification: @NGimhana Do you remember why this change was made? If this change was intentional, I think we can remove the import `HashTable` under imports.

File: webapp/fmprod/src/test/java/org/apache/oodt/cas/product/jaxrs/resources/MetadataResourceTest.java [hunk 11]
```diff
@@ -65,7 +66,7 @@ public void testXmlMarshalling() throws IOException, JAXBException,
       + "</metadata>";
 
     // Create a MetadataResource using a Metadata instance.
-    Hashtable metadataEntries = new Hashtable<String, Object>();
+    HashMap<String, Object> metadataEntries = new HashMap<>();
```
Comment (IMS94) [inline webapp/fmprod/src/test/java/org/apache/oodt/cas/product/jaxrs/resources/MetadataResourceTest.java]: Same. If this change was intentional, we should be able to remove `HashTable` import

File: webapp/fmprod/src/test/java/org/apache/oodt/cas/product/jaxrs_rest/RestAPITest.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    public static void destroy() {
+        server.stop();
+        server.destroy();
+    }
+
+    @Test
+    public void testGetProductResourceWithWebClient() {
+        WebClient client = WebClient.create(ENDPOINT_ADDRESS);
+        ProductResource productResource = client.get(ProductResource.class);
+        assertEquals(0, productResource.getProductId());
```
Comment (IMS94) [inline webapp/fmprod/src/test/java/org/apache/oodt/cas/product/jaxrs_rest/RestAPITest.java]: A suggestion. I assume the `ProductResource` returned here should be empty since we are not ingesting any file prior to fetching that. Isn't it better to add a section to ingest a file first and get the corresponding product here and assert?

File: webapp/fmprod/src/main/webapp/WEB-INF/web.xml (hunk 9 above)
Comment (NGimhana) [reply webapp/fmprod/src/main/webapp/WEB-INF/web.xml]: Yes Indeed. But at the current situation to maintain back-compatibility I separate v2 APIs completely from existing ones. [PR#97](https://github.com/apache/oodt/pull/97/commits/d88b9500d76fa98979ce5213e904f5fa2e03c0fc) fixes this.

File: webapp/fmprod/src/test/java/org/apache/oodt/cas/product/jaxrs/resources/DatasetResourceTest.java (hunk 10 above)
Comment (NGimhana) [reply webapp/fmprod/src/test/java/org/apache/oodt/cas/product/jaxrs/resources/DatasetResourceTest.java]:
```suggestion
 @Deprecated
 public void addMetadata(Hashtable<String, Object> metadata) {
     addMetadata(metadata, false);
 }
```
Above method in [cas/Metadata.java](https://github.com/apache/oodt/blob/master/metadata/src/main/java/org/apache/oodt/cas/metadata/Metadata.java) is deprecated. Just affect to Test cases in this case. So we can remove Hashtable import.

Comment (NGimhana) [reply webapp/fmprod/src/test/java/org/apache/oodt/cas/product/jaxrs/resources/DatasetResourceTest.java]:
```java
@Deprecated
public void addMetadata(Hashtable<String, Object> metadata) {
    addMetadata(metadata, false);
}
```

Above method in [cas/Metadata.java](https://github.com/apache/oodt/blob/master/metadata/src/main/java/org/apache/oodt/cas/metadata/Metadata.java) is deprecated. Just affect to Test cases in this case. So we can remove Hashtable import.

File: webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/services/FileManagerJaxrsServiceV2.java (hunk 8 above)
Comment (NGimhana) [reply webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/services/FileManagerJaxrsServiceV2.java]: Logs used for just testing purposes. Exceptions are thrown as payloads.

File: webapp/fmprod/pom.xml [hunk 13]
```diff
@@ -72,15 +72,23 @@
...
     <dependency>
       <groupId>org.apache.cxf</groupId>
       <artifactId>cxf-rt-transports-local</artifactId>
       <scope>test</scope>
+      <version>3.1.6</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.cxf</groupId>
+      <artifactId>cxf-rt-transports-http-jetty</artifactId>
+      <version>3.1.6</version>
```
[the remark this answers is not in the document]
Comment (NGimhana) [reply webapp/fmprod/pom.xml]: Added that to oodt-core pom.xml

File: webapp/fmprod/src/test/java/org/apache/oodt/cas/product/jaxrs_rest/RestAPITest.java (hunk 12 above)
Comment (IMS94) [reply webapp/fmprod/src/test/java/org/apache/oodt/cas/product/jaxrs_rest/RestAPITest.java]: @NGimhana I think we should ignore this for now as you suggested on the call last week?

Comment (NGimhana) [reply webapp/fmprod/src/test/java/org/apache/oodt/cas/product/jaxrs_rest/RestAPITest.java]: Yes..I think it is safe to ignore this test at the current moment.

File: webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/filters/CORSFilter.java [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+    @Override
+    public void init(FilterConfig filterConfig) throws ServletException {
+
+    }
+
+    @Override
+    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain)
+            throws IOException, ServletException {
+        HttpServletRequest request = (HttpServletRequest) servletRequest;
+        System.out.println("CORSFilter HTTP Request: " + request.getMethod());
```
[the remark this answers is not in the document]
Comment (NGimhana) [reply webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/filters/CORSFilter.java]: Replaced debug with slf4j logger.

File: webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/exceptions/BadRequestException.java [hunk 15]
```diff
@@ -13,33 +13,29 @@
...
+ * This type of exception returns an HTTP 'bad request' response (status code 400) with an
+ * additional message.
+ *
  * @author rlaidlaw
  * @version $Revision$
  */
-public class BadRequestException extends WebApplicationException
-{
-  // Auto-generated ID for serialization.
+public class BadRequestException extends WebApplicationException {
```
Comment (IMS94) [inline webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/exceptions/BadRequestException.java]: Could you check if this class actually has your changes? If no, let's not format such classes.

Comment (NGimhana) [reply webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/exceptions/BadRequestException.java]: Yes, I did some changes to the class. (Changed its constructor a bit)

Comment (IMS94) [reply webapp/fmprod/src/main/java/org/apache/oodt/cas/product/jaxrs/exceptions/BadRequestException.java]: Great. I will merge this once I check this on an IDE.
